base_name_search_improved/models/ir_model.py

```diff
@@ -0,0 +1,264 @@
+        results.extend(rec_ids)
+    return results
+
+
+def patch_name_search():
+    @api.model
+    def _name_search(
+        self,
+        name,
+        domain=None,
+        operator="ilike",
+        limit=None,
+        order=None,
+    ):
+        # Perform standard name search
+        res = _name_search.origin(
+            self,
+            name,
+            domain=domain,
+            operator=operator,
+            limit=limit,
+            order=order,
+        )
+        if name and _get_use_smart_name_search(self.sudo()) and operator in ALLOWED_OPS:
+            # _name_search.origin is a query, we need to convert it to a list
+            res = self.browse(res).ids
+            limit = limit or 0
+
+            # we add domain
+            args = domain
+            args = args or [] + _get_name_search_domain(self.sudo())
+
+            # Support a list of fields to search on
+            all_names = _get_rec_names(self.sudo())
+            base_domain = args or []
+            # Try regular search on each additional search field
+            for rec_name in all_names[1:]:
+                new_domain = [(rec_name, operator, name)]
+                res = _extend_name_results(
+                    self, base_domain + new_domain, res, limit, order
+                )
+            # Try ordered word search on each of the search fields
+            for rec_name in all_names:
+                new_domain = [(rec_name, operator, name.replace(" ", "%"))]
+                res = _extend_name_results(
+                    self, base_domain + new_domain, res, limit, order
+                )
+            # Try unordered word search on each of the search fields
+            # we only perform this search if we have at least one
+            # separator character
+            # also, if have raise the limit we skeep this iteration
+            if " " in name and len(res) < limit:
+                new_domain = []
+                for word in name.split():
+                    word_domain = []
+                    for rec_name in all_names:
+                        word_domain = (
+                            word_domain and ["|"] + word_domain or word_domain
+                        ) + [(rec_name, operator, word)]
+                    new_domain = (
+                        new_domain and ["&"] + new_domain or new_domain
+                    ) + word_domain
+                res = _extend_name_results(
+                    self, base_domain + new_domain, res, limit, order
+                )
+
+        return res
+
+    return _name_search
+
+
+class Base(models.AbstractModel):
+    _inherit = "base"
+
+    # TODO perhaps better to create only the field when enabled on the model
+    smart_search = fields.Char(
+        compute="_compute_smart_search",
+        search="_search_smart_search",
+    )
+
+    def _compute_smart_search(self):
+        self.smart_search = False
+
+    @api.model
+    def _search_smart_search(self, operator, value):
```

Comment:
While it looks like [smart search is included in tests](https://github.com/ACCOMODATA-BVBA/server-tools/blob/f5992d2f224cc0f87b54986c09e6670b3b781825/base_name_search_improved/tests/test_name_search.py#L15), this function does not seem to receive test coverage. Any idea why?